geomstats/stiefel.py

```diff
@@ -66,3 +70,284 @@ def random_uniform(self, n_samples=1):
+        matrix_k = point - gs.matmul(base_point, matrix_m)
+
+        matrix_q = gs.zeros(matrix_k.shape)
+        matrix_n = gs.zeros((n_logs, p, p))
+        for i, k in enumerate(matrix_k):
+            matrix_q[i], matrix_n[i] = gs.linalg.qr(k)
+
+        # Orthogonal completion
+        matrix_w = gs.concatenate([matrix_m, matrix_n], axis=1)
+
+        matrix_v = gs.zeros((
+            matrix_w.shape[0],
+            max(matrix_w.shape[1], matrix_w.shape[2]),
+            max(matrix_w.shape[1], matrix_w.shape[2])
+            ))
+
+        for i, w in enumerate(matrix_w):
+            matrix_v[i], _ = gs.linalg.qr(w, mode="complete")
+
+        # Procrustes preprocessing
+        [matrix_d, matrix_s, matrix_r] = gs.linalg.svd(
+            matrix_v[:, p:2*p, p:2*p])
+
+        matrix_rd = gs.matmul(matrix_r, gs.transpose(matrix_d, axes=(0, 2, 1)))
+        matrix_v[:, :, p:2*p] = gs.matmul(matrix_v[:, :, p:2*p], matrix_rd)
+        matrix_v = gs.concatenate(
+            [gs.concatenate([matrix_m, matrix_n], axis=1),
+             matrix_v[:, :, p:2*p]],
+            axis=2)
+
+        for k in range(max_iter):
+
+            matrix_lv = gs.logm(matrix_v)
+
+            matrix_c = matrix_lv[:, p:2*p, p:2*p]
+            norm_matrix_c = gs.linalg.norm(matrix_c, ord=2, axis=(1, 2))
+
+            if gs.all(norm_matrix_c < tol):
+                # Convergence achieved
+                break
+
+            matrix_phi = gs.expm(-matrix_c)
+            matrix_v[:, :, p:2*p] = gs.matmul(
+                matrix_v[:, :, p:2*p], matrix_phi)
+
+        matrix_xv = gs.matmul(base_point, matrix_lv[:, 0:p, 0:p])
+        matrix_qv = gs.matmul(matrix_q, matrix_lv[:, p:2*p, 0:p])
+
+        return matrix_xv + matrix_qv
+
+    def retraction(self, tangent_vec, base_point):
+        """
+        Retraction map, based on QR-decomposion:
+        P_x(V) = qf(X + V)
+        """
+        tangent_vec = gs.to_ndarray(tangent_vec, to_ndim=3)
+        n_tangent_vecs, _, _ = tangent_vec.shape
+
+        base_point = gs.to_ndarray(base_point, to_ndim=3)
+        n_base_points, n, p = base_point.shape
+
+        assert (n_tangent_vecs == n_base_points
+                or n_tangent_vecs == 1
+                or n_base_points == 1)
+
+        if n_base_points == 1:
+            base_point = gs.tile(base_point, (n_tangent_vecs, 1, 1))
+        if n_tangent_vecs == 1:
+            tangent_vec = gs.tile(tangent_vec, (n_base_points, 1, 1))
+        n_retractions = gs.maximum(n_base_points, n_tangent_vecs)
+        # Q, R = gs.linalg.qr(base_point + tangent_vec)
+        # TODO: remove cycle, when qr will be vectorized
+        matrix_q = gs.zeros_like(base_point)
+        matrix_r = gs.zeros((n_retractions, p, p))
+
+        for i, k in enumerate(base_point + tangent_vec):
+            matrix_q[i], matrix_r[i] = gs.linalg.qr(k)
+
+        # flipping signs
+        # Q = gs.matmul(Q, gs.diag(gs.sign(gs.sign(gs.diagonal(R)) + 0.5)))
```

Comment:
why is this still commented? Either get rid of it or uncomment -

Reply:
Done.